### ISSUE TITLE
Bootstrap go from source

### ISSUE THREAD

**Issue by linsui** [NixOS/nixpkgs#277244]
Go is pretty easy to bootstrap. Currently there are 2 paths to bootstrap it. One is building it with gccgo which is currently used for musl target. Another one is bootstraping it from go 1.4, then go 1.17.3.

Go has switched to [x - 2](https://github.com/golang/go/issues/54265). Currently gccgo supports go 1.18 so when go 1.12 is released the path becomes gccgo -> go 1.20 -> go 1.22. But it's expected that gccgo will also be updated and the path won't be too long.

Could we bootstrap go from source? Is there any other concen?

See related discussion in https://github.com/NixOS/nixpkgs/pull/207681.

@NixOS/golang

**Comment by zowoq** [comment]
We used to bootstrap go from source but that only worked for x86-linux and x86-darwin. https://github.com/NixOS/nixpkgs/pull/113230

I'd rather we just have a simple and consistent bootstrap across platforms we care about instead of needing to maintain cross platform bootstraps, chaining bootstrap versions and having to deal with uploading to tarballs.nixos.

Strong oppose, it isn't worth the hassle.

**Comment by linsui** [comment]
https://github.com/NixOS/nixpkgs/pull/113230 doesn't provide reasons. And we can use gccgo instead.

**Comment by zowoq** [comment]
> https://github.com/NixOS/nixpkgs/pull/113230 doesn't provide reasons.

I've been maintaining `go` in nixpkgs for three years now and I was the one who merged that PR. The reasons may not be explicit on the PR but non-x86 was why the source bootstrap was dropped.

> And we can use gccgo instead.

Isn't supported on darwin and gccgo has the same issues as the source bootstrap anyway.

**Comment by AndersonTorres** [comment]
We have crazy people porting MES to Nixpkgs. I don't believe we should stop anyone interested on bootstrapping on the grounds of "isn't worth the hassle".

Certainly it is way more practical to use precompiled binaries. However, ideally we should provide a from-source bootstrap rout for those interested, and also for testing purposes.

**Comment by linsui** [comment]
> Isn't supported on darwin and gccgo has the same issues as the source bootstrap anyway.

We can bootstrap supported platform only. The effort is near to zero.

**Comment by zowoq** [comment]
> We can bootstrap supported platform only.

As I said previously, I prefer a consistent bootstrap.

> The effort is near to zero.

It'll still need to be maintained and it'll likely be me who has to do it.

**Comment by Aleksanaa** [comment]
I don't think we necessarily need to use consistent bootstrap on different platforms. At least we don't do this in C. If a bootstrap from source proves to work (and doesn't break regularly), there's no reason why we shouldn't accept it.

I can try something related after finishing my final exam this week.

**Comment by Aleksanaa** [comment]
I don't think we necessarily need to use consistent bootstrap on different platforms. At least we don't do this in C. If a bootstrap from source proves to work (and doesn't break regularly), there's no reason why we shouldn't accept it.

I can try something related after finishing my final exam this week.

**Comment by linsui** [comment]
Gccgo is already used to bootstrap some target. You just need to this line to bootstrap more target.

https://github.com/NixOS/nixpkgs/blob/4de6500caffe0c0e6a113b82f3382a6e58a4a515/pkgs/development/compilers/go/1.21.nix#L20

**Comment by doronbehar** [comment]
> > And we can use gccgo instead.
> 
> Isn't supported on darwin and gccgo has the same issues as the source bootstrap anyway.

@zowoq just out of interest, what were the issues with gccgo on Darwin?

**Comment by zowoq** [comment]
> @zowoq just out of interest, what were the issues with gccgo on Darwin?

It has never worked:

> https://gcc.gnu.org/bugzilla/show_bug.cgi?id=46986
> Go is not supported on Darwin
> Reported: 2010-12-17

**Comment by raboof** [comment]
As an unscientific data point (from running with `allowNonSource = false` for a while), `go` seems to be one of the most commonly found dependencies not built from source. I think it would be valuable to do the source bootstrap on platforms where that is possible.

At the same time I appreciate that this will increase complexity and maintainer burden, and as long as the current maintainer team does not feel comfortable with it we shouldn't do it. I just wanted to share that I would consider it valuable if at some point we do find the bandwidth ;) .

**Comment by AndersonTorres** [comment]
The complexity of keeping bootstrap routes is heinous. I can understand the concern.
What about creating a team dedicated to this specific task?
